### DISCUSSION TITLE
Execute each table row as a scenario when passing table as a argument

### DISCUSSION THREAD

**Discussion by guntribam** [getgauge/gauge#1978] [General]
I am testing Gauge and need to clarify this... Currently i am using tables in 2 ways in my test:
One is like this:

This is my ```example.spec```

## Lançamentos

| Valor     | Descrição   | conta  | Tipo de Lançamento | data       |
|-----------|-------------|--------|--------------------|------------|
| tresMil   | geladeira   | cartao | débito             | 21/05/2021 |
| quatroMil | lava e seca | cartao | crédito            | 22/05/2021 |
| quatroMil | geladeira   | cartao | crédito            | 23/05/2021 |

* Cadastrar Lançamentos

| Valor   | Descrição   | conta   | Tipo de Lançamento   | data   |
|---------|-------------|---------|----------------------|--------|
| <Valor*>  | <Descrição*> | <conta*> | <Tipo de Lançamento*> | <data*> |

forget about these asterisks(*) i've put them only to render the markdown, they don't exist in the code

I this first example i am using 2 tables and passing the second one to my step implementation. When i run this spec this is the result:

![Captura de tela de 2021-04-26 20-28-36](https://user-images.githubusercontent.com/7470516/116163374-06a6ca80-a6ce-11eb-8510-7f6a7db7a670.png)

This is good, mainly because if the first table scenario fails(and only) i get a good separation like this:

![Captura de tela de 2021-04-26 20-34-16](https://user-images.githubusercontent.com/7470516/116163825-e1668c00-a6ce-11eb-9bd2-e45483079e54.png)

Ok, now it is time for the second example

This is my ```example.spec```
## Lançamentos

* Cadastrar Lançamentos

| Valor     | Descrição   | conta  | Tipo de Lançamento | data       |
|-----------|-------------|--------|--------------------|------------|
| tresMil   | geladeira   | cartao | débito             | 21/05/2021 |
| quatroMil | lava e seca | cartao | crédito            | 22/05/2021 |
| quatroMil | geladeira   | cartao | crédito            | 23/05/2021 |

I'm passing the "full" table as a parameter to my setp with the values that i'll use

In my step i translate the values comming from the markdown to values that my app understand in the function ```"transformTableValues"```

```js
step("Cadastrar Lançamentos <table>", async table => {
  table.entries(async entry => {
    const { text } = await supertest(BASE_URL)
    .post('/api/lancamentos')
    .send(transformTableValues(entry))
    .expect(200)
  })
})
```
Right here is where my problems starts. Look to the report:

![Captura de tela de 2021-04-26 20-46-31](https://user-images.githubusercontent.com/7470516/116164649-833aa880-a6d0-11eb-852b-34e7bfdbb35c.png)

The whole thing became just one scenario when it is supposed to have 3 scenarios. If i have an error on row 2 the whole scenario will show the error, instead of only row 2...

There is a way to separate the scenarios(rows) when passing a table as a inline argument to a step?

**Comment by sriv** [comment]
If you pass the full table to a step, the table is treated as a parameter and not used as a source for the scenario. Hence you get one scenario with a step that uses a table. Your step implementation is hidden from the scenario, so gauge cannot infer the semantics here.

If you need each row to be reported as separate scenario/spec you'll have to use table driven execution. 

So, to your question:

> There is a way to separate the scenarios(rows) when passing a table as a inline argument to a step?

The short answer is no. Try modelling your scenario to be table driven instead.

**Reply by guntribam** [reply]
Thanks for the fast feedback

**Reply by guntribam** [reply]
What is your opinion about the situation explained above become a feature on Gauge:

- Pass a table as a parameter and execute each row as a different scenario. (initial implementation)

- Show the table as the last image on my explanation but using the "green" color for each row that pass the test and "red" color in each row that fails the test with a good separation of the error messages for each scenario. (evolution of the initial implementation)

I'm thinking about something like this (this is concordion)

![image](https://user-images.githubusercontent.com/7470516/116471844-87d79c00-a84b-11eb-9da4-71b4efc1bdf5.png)

In this example you have cell level assertion. I'm not talking about this level of assertion, we could start doing row level assertion

I know nothing about Go, but i don't know .... maybe i can help in some way and learn something new
@sriv
